partiql-value/src/datum.rs

```diff
@@ -0,0 +1,255 @@
+use crate::{Bag, BindingsName, List, Tuple, Value};
+use std::borrow::Cow;
+use std::error::Error;
+
+pub type DatumLowerError = Box<dyn Error>;
+pub type DatumLowerResult<T> = Result<T, DatumLowerError>;
+
+pub trait Datum<D>
+where
+    D: Datum<D>,
+{
+    #[inline]
+    /// Returns true if and only if Value is to be interpreted as `NULL`
+    #[must_use]
+    fn is_null(&self) -> bool {
+        false
+    }
+
+    #[inline]
+    /// Returns true if and only if Value is to be interpreted as `MISSING`
+    #[must_use]
+    fn is_missing(&self) -> bool {
+        false
+    }
+
+    #[inline]
+    /// Returns true if and only if Value is null or missing
+    #[must_use]
+    fn is_absent(&self) -> bool {
+        self.is_null() || self.is_missing()
+    }
+
+    /// Returns true if and only if Value is an integer, real, or decimal
+    #[must_use]
+    fn is_number(&self) -> bool;
+
+    #[inline]
+    /// Returns true if Value is neither null nor missing
+    #[must_use]
+    fn is_present(&self) -> bool {
+        !self.is_absent()
+    }
+
+    #[must_use]
+    fn is_sequence(&self) -> bool;
+
+    #[must_use]
+    fn is_ordered(&self) -> bool;
+}
+
+pub trait DatumValue<D>: Clone + Datum<D>
+where
+    D: Datum<D>,
+{
+    fn into_lower(self) -> DatumLowerResult<D>;
+}
+
+pub trait DatumCategory<'a> {
+    fn category(&'a self) -> DatumCategoryRef<'a>;
+    fn into_category(self) -> DatumCategoryOwned;
+}
+
+#[derive(Debug)]
+pub enum DatumCategoryRef<'a> {
+    Null,
+    Missing,
+    Tuple(DatumTupleRef<'a>),
+    Sequence(DatumSeqRef<'a>),
+    Scalar(DatumValueRef<'a>),
+}
+
+#[derive(Debug)]
+pub enum DatumCategoryOwned {
```

Comment:
Curious about the structuring of these categories. PLK's Datum stores the Datum's kind/code as a PType https://github.com/partiql/partiql-lang-kotlin/blob/a7b772e4102fb4a0d61e2280202ed38aca269469/partiql-spi/src/main/java/org/partiql/spi/value/Datum.java#L47-L51 whereas this approach has Datum defining its own enum.

Reply:
For one, the typing in PLK is further along than here. Perhaps, partiql-types will grow to take over some of this.

Using the `PType` approach would also require large blocks of `if`/`else if`/etc. for the cases in which this is used. `enum`s and `match` are the more idiomatic approach in rust.

Also... `PType` is also a bit odd. Decimals are `PTypeDecimal`, Text types are `PTypeWithMaxLength`, and other 'scalar's are `PTypePrimitive`. It's difficult with `PType` to split values into Scalar vs. Sequence vs. Tuple cleanly.